### PR TITLE
Conv2dGenerator generates dummy kernel when kernel_id != 0

### PR DESCRIPTION
**Pull request by sjw36**
Initial support in C-API to generate sub-kernel.

* the mlir-miopen-driver will need to call the generator with successive kernel_id

### CODE REVIEW

File: mlir/lib/Dialect/MIOpen/Generator/Conv2dGenerator.cpp [hunk 1]
```diff
@@ -51,7 +51,7 @@ int Conv2dGenerator::getKernelCount(const char *arguments) {
 
   auto operation = argMap["operation"];
   if (operation == "conv2d") {
-    count = 1;
+    count = 4;
```
Comment (whchung) [inline mlir/lib/Dialect/MIOpen/Generator/Conv2dGenerator.cpp]: please improve the logic so it counts based on `dilation` and `stride`. check with @asroy on the actual formula.

File: mlir/test/Dialect/MIOpen/lowering_affix_params.mlir [hunk 2]
```diff
@@ -2,4 +2,4 @@
 // RUN: mlir-miopen-driver -p --operation=conv2d_dummy -miopen-lowering -miopen-affine-transform -miopen-affix-params | FileCheck %s
 
 // CHECK: module {{.*}}
-// CHECK-NEXT: func @{{.*}}(%{{.*}}: memref<{{.*}}>, %{{.*}}: memref<{{.*}}>, %arg2: memref<{{.*}}>) attributes {block_size = {{.*}} : i32, grid_size = {{.*}} : i32, kernel}
+// CHECK-NEXT: func @{{.*}}(%{{.*}}: memref<{{.*}}>, %{{.*}}: memref<{{.*}}>, %arg2: memref<{{.*}}>) attributes {block_size = {{.*}} : i32, grid_size = {{.*}} : i32, kernel = 0 : i32}
```
Comment (whchung) [inline mlir/test/Dialect/MIOpen/lowering_affix_params.mlir:5]: please supply unit test where `kernel_id != 0` and we see `miopen.conv2d_dummy` gets populated. probably it needs to be a separate test file.

**Review by whchung — DISMISSED** [review]

File: mlir/lib/Dialect/MIOpen/Generator/Conv2dGenerator.cpp (hunk 1 above)
Comment (sjw36) [reply mlir/lib/Dialect/MIOpen/Generator/Conv2dGenerator.cpp]: Yes, this was a placeholder. @asroy what is the formula?

File: mlir/test/mlir-miopen-driver/populate_subkernels.mlir [hunk 3]
```diff
@@ -0,0 +1,16 @@
+// RUN: mlir-miopen-driver -conv-config "--operation conv2d --arch gfx906 --num_cu 64 --fil_layout GNCHW --in_type fp32 --fil_type fp32 --out_type fp32 --in_layout NGCHW --out_layout NGCHW --batchsize 64 --in_channels 1024 --out_channels 1024 --in_h 14 --in_w 14 --out_h 14 --out_w 14 --fil_h 1 --fil_w 1 --dilation_h 1 --dilation_w 1 --conv_stride_h 1 --conv_stride_w 1 --padding_h 0 --padding_w 0 --kernel_name conv2d_fwd --groupsize 1 --kernel_id 1" | FileCheck %s --check-prefix=KERNEL1
```
Comment (whchung) [inline mlir/test/mlir-miopen-driver/populate_subkernels.mlir:2]: how about add one more line with `kernel_id 0`?

and perhaps add some comments in this file to remind ourselves that all of these shall be changed when `miopen.conv2d_bwd_data` implementation is ready.

File: mlir/lib/Dialect/MIOpen/Generator/Conv2dGenerator.cpp (hunk 1 above)
Comment (whchung) [reply mlir/lib/Dialect/MIOpen/Generator/Conv2dGenerator.cpp]: + @asleepzzz for awareness. please supply the formula to correctly count the number of kernels after the split.

File: mlir/tools/mlir-miopen-driver/mlir-miopen-driver.cpp [hunk 4]
```diff
@@ -2235,9 +2241,12 @@ int main(int argc, char **argv) {
   }
 
   // Apply passes.
-  if (failed(runMLIRPasses(module, passPipeline, genConfig.kernelName))) {
-    llvm::errs() << "Lowering failed.\n";
-    exit(1);
+  for (const auto &kernelName : kernels) {
+    std::cout << "Kernel - " << kernelName << std::endl;
```
Comment (whchung) [inline mlir/tools/mlir-miopen-driver/mlir-miopen-driver.cpp]: This seems like a debug logic. Please remove it.

File: mlir/tools/mlir-miopen-driver/mlir-miopen-driver.cpp [hunk 5]
```diff
@@ -2194,25 +2195,30 @@ int main(int argc, char **argv) {
...
-  if (dataTypeStr == "f32") {
-    dataType = builder.getF32Type();
-  } else if (dataTypeStr == "f16") {
-    dataType = builder.getF16Type();
-  } else if (dataTypeStr == "bf16") {
-    dataType = builder.getIntegerType(16);
-  }
+  const auto &genConfig = conv2dGenerator.getConfig();
 
+  std::vector<std::string> kernels;
```
Comment (whchung) [inline mlir/tools/mlir-miopen-driver/mlir-miopen-driver.cpp]: Can we use SmallVector<StringRef>?

File: mlir/tools/mlir-miopen-driver/mlir-miopen-driver.cpp [hunk 6]
```diff
@@ -37,6 +37,7 @@
 
 #include "bf16convert.hpp"
 #include "mlir/Dialect/SCF/EDSC/Builders.h"
+#include <iostream>
```
Comment (whchung) [inline mlir/tools/mlir-miopen-driver/mlir-miopen-driver.cpp]: I think this is only used for debugging?

**Comment by sjw36** [comment]
@whchung @jerryyin , I am cleaning up and MIOpenToGPU conversion to look for all FuncOp's with kernel attribute, instead of passing the kernel-names. Do you guys need that feature?

**Comment by sjw36** [comment]
Also automatic creation of wrapper GPUModule (instead of finding one and putting a kernel in it..).

**Comment by whchung** [comment]
> @whchung @jerryyin , I am cleaning up and MIOpenToGPU conversion to look for all FuncOp's with kernel attribute, instead of passing the kernel-names. Do you guys need that feature?

I think we still need this because we may not need to lower every FuncOp with kernel attribute. In your new scheme can we still restrict the FuncOp be lowered?

**Comment by whchung** [comment]
> Also automatic creation of wrapper GPUModule (instead of finding one and putting a kernel in it..).

We should be able to get rid of this, provided you fix relevant unit tests ;)

**Comment by sjw36** [comment]
> > @whchung @jerryyin , I am cleaning up and MIOpenToGPU conversion to look for all FuncOp's with kernel attribute, instead of passing the kernel-names. Do you guys need that feature?
> 
> I think we still need this because we may not need to lower every FuncOp with kernel attribute. In your new scheme can we still restrict the FuncOp be lowered?

Nope, just converting based on the kernel attribute. What scenarios do we need to limit?

**Comment by whchung** [comment]
> > > @whchung @jerryyin , I am cleaning up and MIOpenToGPU conversion to look for all FuncOp's with kernel attribute, instead of passing the kernel-names. Do you guys need that feature?
> > 
> > 
> > I think we still need this because we may not need to lower every FuncOp with kernel attribute. In your new scheme can we still restrict the FuncOp be lowered?
> 
> Nope, just converting based on the kernel attribute. What scenarios do we need to limit?

When we spit a kernel into multiple ones we don’t really need to go through the lowering process for each of them, instead we could use the name to specify which gets lowered. Thats the idea behind my experimental PR which splits conv2d, and why we put that splitting pass in the very beginning of the flow.

In case we can devise another way to achieve that I’m all for it

**Comment by sjw36** [comment]
They would all need conversion to GPU for execution right?
If we really need to avoid conversion for some of these kernels we could base that on the kernel attribute value (as you can see below I have them enumerated) or add another attribute.

These are the kernel functions the driver is now generating:

```
 module  {
   func @miopen_conv2d_gkcyx_ngchw_ngkhw(%arg0: memref<1x128x8x3x3xf32>, %arg1: memref<128x1x8x32x32xf32>, %arg2: memref<128x1x128x30x30xf32>) attributes {kernel = 0 : i32} {
     miopen.conv2d(%arg0, %arg1, %arg2) {arch = "gfx906", dilations = [1 : i32, 1 : i32], filter_layout = ["g", "k", "c", "y", "x"], input_layout = ["ni", "gi", "ci", "hi", "wi"], num_cu = 64 : i32, output_layout = ["no", "go", "ko", "ho", "wo"], padding = [0 : i32, 0 : i32], strides = [1 : i32, 1 : i32]} : memref<1x128x8x3x3xf32>, memref<128x1x8x32x32xf32>, memref<128x1x128x30x30xf32>
     return
   }
   func @miopen_conv2d_gkcyx_ngchw_ngkhw_1(%arg0: memref<1x128x8x3x3xf32>, %arg1: memref<128x1x8x32x32xf32>, %arg2: memref<128x1x128x30x30xf32>) attributes {kernel = 1 : i32} {
     miopen.conv2d_dummy(%arg0, %arg1, %arg2) {arch = "gfx906", dilations = [1 : i32, 1 : i32], filter_layout = ["g", "k", "c", "y", "x"], input_layout = ["ni", "gi", "ci", "hi", "wi"], num_cu = 64 : i32, output_layout = ["no", "go", "ko", "ho", "wo"], padding = [0 : i32, 0 : i32], strides = [1 : i32, 1 : i32]} : memref<1x128x8x3x3xf32>, memref<128x1x8x32x32xf32>, memref<128x1x128x30x30xf32>
     return
   }
   func @miopen_conv2d_gkcyx_ngchw_ngkhw_2(%arg0: memref<1x128x8x3x3xf32>, %arg1: memref<128x1x8x32x32xf32>, %arg2: memref<128x1x128x30x30xf32>) attributes {kernel = 2 : i32} {
     miopen.conv2d_dummy(%arg0, %arg1, %arg2) {arch = "gfx906", dilations = [1 : i32, 1 : i32], filter_layout = ["g", "k", "c", "y", "x"], input_layout = ["ni", "gi", "ci", "hi", "wi"], num_cu = 64 : i32, output_layout = ["no", "go", "ko", "ho", "wo"], padding = [0 : i32, 0 : i32], strides = [1 : i32, 1 : i32]} : memref<1x128x8x3x3xf32>, memref<128x1x8x32x32xf32>, memref<128x1x128x30x30xf32>
     return
   }
   func @miopen_conv2d_gkcyx_ngchw_ngkhw_3(%arg0: memref<1x128x8x3x3xf32>, %arg1: memref<128x1x8x32x32xf32>, %arg2: memref<128x1x128x30x30xf32>) attributes {kernel = 3 : i32} {
     miopen.conv2d_dummy(%arg0, %arg1, %arg2) {arch = "gfx906", dilations = [1 : i32, 1 : i32], filter_layout = ["g", "k", "c", "y", "x"], input_layout = ["ni", "gi", "ci", "hi", "wi"], num_cu = 64 : i32, output_layout = ["no", "go", "ko", "ho", "wo"], padding = [0 : i32, 0 : i32], strides = [1 : i32, 1 : i32]} : memref<1x128x8x3x3xf32>, memref<128x1x8x32x32xf32>, memref<128x1x128x30x30xf32>
     return
   }
 }
```

After MIOpenToGPU there are 4 gpu::modules with 1 gpu::func each: 

```
 module attributes {gpu.container_module}  {
   gpu.module @miopen_conv2d_gkcyx_ngchw_ngkhw_module {
     gpu.func @miopen_conv2d_gkcyx_ngchw_ngkhw(%arg0: memref<1x128x8x3x3xf32>, %arg1: memref<128x1x8x32x32xf32>, %arg2: memref<128x1x128x30x30xf32>) workgroup(%arg3 : memref<2048xf32, 3>) private(%arg4 : memref<1x1x8xf32, 5>, %arg5 : memref<1x1x8xf32, 5>, %arg6 : memref<1x1x8xf32, 5>, %arg7 : memref<1x1x8xf32, 5>, %arg8 : memref<3xi32, 5>, %arg9 : memref<3xi32, 5>, %arg10 : memref<3xi32, 5>, %arg11 : memref<3xi32, 5>, %arg12 : memref<1x4x1xf32, 5>, %arg13 : memref<1x1x4xf32, 5>, %arg14 : memref<1x8x8xf32, 5>) kernel attributes {block_size = 256 : i32, grid_size = 900 : i32} {
       %c9 = constant 9 : index
       %c3 = constant 3 : index
       %c1024 = constant 1024 : index
       %c-1 = constant -1 : index
       %c1 = constant 1 : index
       br ^bb1
     ^bb1:  // pred: ^bb0
       %0 = "gpu.block_id"() {dimension = "x"} : () -> index
       %1 = divi_signed %0, %c900 : index
       %2 = remi_signed %0, %c900 : index
       ...
       gpu::return
     }
   } 
   gpu.module @miopen_conv2d_gkcyx_ngchw_ngkhw_1_module {
     gpu.func @miopen_conv2d_gkcyx_ngchw_ngkhw_1(%arg0: memref<1x128x8x3x3xf32>, %arg1: memref<128x1x8x32x32xf32>, %arg2: memref<128x1x128x30x30xf32>) kernel attributes {block_size = 1 : i32, grid_size = 1 : i32} {
       br ^bb1
     ^bb1:  // pred: ^bb0
       gpu.return
     }
   }
   gpu.module @miopen_conv2d_gkcyx_ngchw_ngkhw_2_module {
     gpu.func @miopen_conv2d_gkcyx_ngchw_ngkhw_2(%arg0: memref<1x128x8x3x3xf32>, %arg1: memref<128x1x8x32x32xf32>, %arg2: memref<128x1x128x30x30xf32>) kernel attributes {block_size = 1 : i32, grid_size = 1 : i32} {
       br ^bb1
     ^bb1:  // pred: ^bb0
       gpu.return
     }
   }
   gpu.module @miopen_conv2d_gkcyx_ngchw_ngkhw_3_module {
     gpu.func @miopen_conv2d_gkcyx_ngchw_ngkhw_3(%arg0: memref<1x128x8x3x3xf32>, %arg1: memref<128x1x8x32x32xf32>, %arg2: memref<128x1x128x30x30xf32>) kernel attributes {block_size = 1 : i32, grid_size = 1 : i32} {
       br ^bb1
     ^bb1:  // pred: ^bb0
       gpu.return
     }
   }  
```

**Comment by whchung** [comment]
@sjw36 I think this could be a question to @jerryyin : do we need to lower all `FuncOp` instances with `kernel` attributes in one shot, and let the clients pick the compiled results one by one, or we lower them piecewise?

**Comment by sjw36** [comment]
The API should now allow building each kernel individually (piecewise) so it would be up to the client to decide.

**Comment by jerryyin** [comment]
> do we need to lower all FuncOp instances with kernel attributes in one shot, and let the clients pick the compiled results one by one, or we lower them piecewise?

Specifically for my preference, the way MIOpen generate a kernel each time is that it has to lower from the very beginning to the very end. So from the efficiency point of view, I'd like the client to lower them piecewise, one piece at a time. For example, when the client choose the first kernel, we only lower the first kernel and not lower everything else. Similarly, when the second time client choose another kernel, we only lowering that kernel specified by client.

On the other hand, for lowering everything in one shot and let client pick result one by one, it is still functional but does redundant work each time. For example, when the client choose the first kernel, we lowering not only the first kernel but also all other kernels, the client will pick the first kernel and ignore those other kernels. When the client choose another kernel, we still get this pattern of doing more work than client needed.

**Comment by sjw36** [comment]
With the new kernel_id selection field, the client would specify kernel_id=1 and we would generate:

```
   func @miopen_conv2d_gkcyx_ngchw_ngkhw_1(%arg0: memref<1x128x8x3x3xf32>, %arg1: memref<128x1x8x32x32xf32>, %arg2: memref<128x1x128x30x30xf32>) attributes {kernel = 1 : i32} {
     miopen.conv2d_dummy(%arg0, %arg1, %arg2) {arch = "gfx906", dilations = [1 : i32, 1 : i32], filter_layout = ["g", "k", "c", "y", "x"], input_layout = ["ni", "gi", "ci", "hi", "wi"], num_cu = 64 : i32, output_layout = ["no", "go", "ko", "ho", "wo"], padding = [0 : i32, 0 : i32], strides = [1 : i32, 1 : i32]} : memref<1x128x8x3x3xf32>, memref<128x1x8x32x32xf32>, memref<128x1x128x30x30xf32>
     return
   }
```

So that would be the only kernel that would be compiled.

**Comment by jerryyin** [comment]
@sjw36 

> So that would be the only kernel that would be compiled.

Sounds good to me, thanks. If I understand it right, this approach is used only in the prototype? When Kevin and Letao finished their work in populating non-dummy kernels, is it going to be your or Jack's approach in kernel population?

**Comment by sjw36** [comment]
Yes, the kernel generator should populate the appropriate kernel based on the "operation" type and "kernel_id".

**Comment by jerryyin** [comment]
@sjw36 Let me know when the branch is ready so I will take it to do the development of MIOpen with the branch.

**Comment by sjw36** [comment]
> @sjw36 Let me know when the branch is ready so I will take it to do the development of MIOpen with the branch.

Current should work. I have a few cleanup changes and working on fixing the tests.

**Comment by jerryyin** [comment]
@sjw36 If you do not seek for merge of this PR and if there's no functional breakage of tests then feel free to ignore those. I don't see it necessary to further the work if we ended up using Jack's approach in #139.

**Comment by whchung** [comment]
> @sjw36 If you do not seek for merge of this PR and if there's no functional breakage of tests then feel free to ignore those. I don't see it necessary to further the work if we ended up using Jack's approach in #139.

Please correct me if I'm wrong here. I think the approach I did in #139 is orthogonal to @sjw36 's work here? #139 provides a end-to-end flow to show how things work in `mlir-miopen-driver` -> `mlir-rocm-runner` fashion. But @sjw36 's work here take client application (such as MIOpen) into perspective. So I believe the intention here is to improve and merge this PR?

**Comment by sjw36** [comment]
@jerryyin I meant this current PR. I believe it will fulfil the needs of the clients.

**Comment by jerryyin** [comment]
We synced offline.

@whchung PR #139 show case the concept. Once Simon's current PR is merged, we'd no longer need the split pass to populate dummy ops.

@sjw36 will continue to pursue the merge of this current PR.

File: mlir/lib/Dialect/MIOpen/Generator/Conv2dGenerator.cpp (hunk 1 above)
Comment (sjw36) [reply mlir/lib/Dialect/MIOpen/Generator/Conv2dGenerator.cpp]: Still haven't seen this but it can be filled in when the sub-kernels are ready.

**Comment by sjw36** [comment]
All tests passing now. Please complete review.

File: mlir/lib/Dialect/MIOpen/Generator/Conv2dGenerator.cpp (hunk 1 above)
Comment (whchung) [reply mlir/lib/Dialect/MIOpen/Generator/Conv2dGenerator.cpp]: It's a long weekend in Asia so probably @asleepzzz and @ltqin are not back in the office yet.

**Comment by asleepzzz** [comment]
the equation:
    const auto gcd_stride_dilation_h = gcd(conv_stride_h, conv_dilation_h);
    const auto gcd_stride_dilation_w = gcd(conv_stride_w, conv_dilation_w);
    const auto ytilda = conv_stride_h / gcd_stride_dilation_h;
    const auto xtilda = conv_stride_w / gcd_stride_dilation_w;
    num_of_kernels = ytilda * xtilda;

please use it for getKernelCount

**Comment by sjw36** [comment]
> the equation:
> const auto gcd_stride_dilation_h = gcd(conv_stride_h, conv_dilation_h);
> const auto gcd_stride_dilation_w = gcd(conv_stride_w, conv_dilation_w);
> const auto ytilda = conv_stride_h / gcd_stride_dilation_h;
> const auto xtilda = conv_stride_w / gcd_stride_dilation_w;
> num_of_kernels = ytilda * xtilda;
> 
> please use it for getKernelCount

Will that be for conv2d or conv2d_bwd_data or conv2d_bwd_weight? 

@whchung since this PR will only generate extra dummy kernels, how about we update getKernelCount when the new kernels are ready. The functionality will be tested with the API changes here (mlir/test/mlir-miopen-driver/populate_subkernels.mlir).

**Comment by whchung** [comment]
> Will that be for conv2d or conv2d_bwd_data or conv2d_bwd_weight?

only for `conv2d_bwd_data`.

> @whchung since this PR will only generate extra dummy kernels, how about we update getKernelCount when the new kernels are ready. The functionality will be tested with the API changes here (mlir/test/mlir-miopen-driver/populate_subkernels.mlir).

yes. we can defer this feature to a subsequent PR.

**Comment by jerryyin** [comment]
Could you expose `miirGetKernelCount()` In the header? I can't find it in `Miir.h`. I'd prefer it to accept a handle to be consistent with the rest of API.

**Comment by sjw36** [comment]
> I'd prefer it to accept a handle to be consistent with the rest of API.

Ok, the client can create the first handle and get the count from that, then generate subsequent handles with subsequent kernel_id's.

**Comment by sjw36** [comment]
@whchung please close the 'change request'.

@jerryyin, update miirGetKernelCount to take handle.

**Review by jerryyin — APPROVED** [review]
LGTM

**Pull request merged**